tools/debian/first_package_finder.py

```diff
@@ -0,0 +1,129 @@
+"""Functions to help find the first package version for a specific release"""
+import urllib.error
+from datetime import datetime, timedelta
+import json
+import gzip
+from urllib import request
+
+import pandas as pd
+
+DEBIAN_RELEASE_VERSIONS_URL = \
+  'https://debian.pages.debian.net/distro-info-data/debian.csv'
+DEBIAN_SNAPSHOT_URL = 'https://snapshot.debian.org/archive/debian/{date}/dists/'
+# `.gz` format always exist for all snapshots
+DEBIAN_SOURCES_URL_EXTENSION = '{version}/main/source/Sources.gz'
+
+# List of ignored versions, mostly too early to be in snapshots
+IGNORED_DEBIAN_VERSIONS = ['experimental', 'buzz', 'rex', 'bo', 'hamm', 'slink', 'potato']
+
+def get_debian_dists_url(date: datetime):
+  """Create an url for snapshot.debian.org to get distribution"""
+  formatted_date = convert_datetime_to_str_datetime(date)
+  return DEBIAN_SNAPSHOT_URL.format(date=formatted_date)
+
+
+def get_debian_sources_url(date: datetime, version: str):
+  """Create an url for snapshot.debian.org"""
+
+  formatted_date = convert_datetime_to_str_datetime(date)
+
+  return DEBIAN_SNAPSHOT_URL.format(
+      date=formatted_date) + DEBIAN_SOURCES_URL_EXTENSION.format(
+          version=version)
+
+
+def convert_datetime_to_str_datetime(input_datetime: datetime) -> str:
+  """Convert datetime object to debian snapshot url string"""
+  return input_datetime.isoformat().replace('-', '').replace(':', '') + 'Z'
+
+
+def retrieve_codename_to_version() -> pd.DataFrame:
+  """Returns the codename to version mapping"""
+  with request.urlopen(DEBIAN_RELEASE_VERSIONS_URL) as csv:
+    df = pd.read_csv(csv, dtype=str)
+    # `series` appears to be `codename` but with no caps
+    df['sources'] = ''
+    df['first seen'] = ''
+    codename_to_version = df.set_index('series')
+    codename_to_version.loc['sid']['version'] = 'unstable'
+
+  return codename_to_version
+
+
+def parse_created_dates_and_set_time(date: str) -> datetime:
+  """Parse created date in debian version csv to datetime plus one day"""
+  result = datetime.strptime(date, '%Y-%m-%d') + timedelta(days=1)
+  # Set minimum date to first debian snapshot
+  return max(result, datetime(2005, 3, 12))
+
+
+def load_sources(date: datetime, dist: str) -> dict[str, str]:
+  """Load the sources file and store in a dictionary of {name: version}"""
+  with request.urlopen(get_debian_sources_url(date, dist)) as res:
+    decompressed = gzip.decompress(res.read()).decode('utf-8', errors='ignore')
```

Comment:
errors='ignore' seems dangerous. did you notice any issues without this specified?

Reply:
![image](https://user-images.githubusercontent.com/106129829/176360042-5d348508-18cb-4bf1-aa61-857a608333f8.png)
Some maintainer names are not proper unicode in the earlier snapshots. They seem to have fixed it in later versions for example:
![image](https://user-images.githubusercontent.com/106129829/176360927-7fcdf61b-d59a-4865-a198-b9e252195504.png)

Reply:
ack!